File: backend/models/BorrowRecord.js
import mongoose from 'mongoose';

const borrowRecordSchema = new mongoose.Schema({
  user: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: [true, 'User is required'],
  },
  book: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Book',
    required: [true, 'Book is required'],
  },
  borrowedAt: {
    type: Date,
    default: Date.now,
  },
  dueAt: {
    type: Date,
    required: [true, 'Due date is required'],
    default: function() {
      // Default to 14 days from borrow date
      const dueDate = new Date();
      dueDate.setDate(dueDate.getDate() + 14);
      return dueDate;
    },
  },
  returnedAt: {
    type: Date,
  },
  renewedCount: {
    type: Number,
    default: 0,
    max: [2, 'Maximum 2 renewals allowed'],
  },
  status: {
    type: String,
    enum: ['borrowed', 'returned', 'overdue', 'renewed'],
    default: 'borrowed',
  },
  fineAmount: {
    type: Number,
    default: 0,
    min: [0, 'Fine amount cannot be negative'],
  },
  finePaid: {
    type: Boolean,
    default: false,
  },
  notes: {
    type: String,
    trim: true,
    maxlength: [500, 'Notes cannot exceed 500 characters'],
  },
}, {
  timestamps: true,
});

// Indexes
borrowRecordSchema.index({ user: 1, status: 1 });
borrowRecordSchema.index({ book: 1, status: 1 });
borrowRecordSchema.index({ dueAt: 1, status: 1 });
borrowRecordSchema.index({ borrowedAt: 1 });

// Virtual to check if overdue
borrowRecordSchema.virtual('isOverdue').get(function() {
  return this.status === 'borrowed' && new Date() > this.dueAt;
});

// Pre-save middleware to update status
borrowRecordSchema.pre('save', function(next) {
  // Update status based on return date and due date
  if (this.returnedAt) {
    this.status = 'returned';
  } else if (new Date() > this.dueAt && this.status === 'borrowed') {
    this.status = 'overdue';
    // Calculate fine (e.g., $1 per day overdue)
    const daysOverdue = Math.ceil((new Date() - this.dueAt) / (1000 * 60 * 60 * 24));
    this.fineAmount = daysOverdue * 1; // $1 per day
  }
  next();
});

// Static method to get active borrows for a user
borrowRecordSchema.statics.getActiveBorrows = function(userId) {
  return this.find({
    user: userId,
    status: { $in: ['borrowed', 'overdue', 'renewed'] },
  }).populate('book');
};

// Static method to get borrow history for a user
borrowRecordSchema.statics.getBorrowHistory = function(userId, options = {}) {
  const { page = 1, limit = 10 } = options;
  
  return this.find({ user: userId })
    .populate('book')
    .sort({ borrowedAt: -1 })
    .skip((page - 1) * limit)
    .limit(limit);
};

// Method to renew the borrow
borrowRecordSchema.methods.renew = function() {
  if (this.renewedCount >= 2) {
    throw new Error('Maximum renewals exceeded');
  }
  
  if (this.status !== 'borrowed') {
    throw new Error('Can only renew active borrows');
  }
  
  // Extend due date by 14 days
  this.dueAt = new Date(this.dueAt.getTime() + (14 * 24 * 60 * 60 * 1000));
  this.renewedCount += 1;
  this.status = 'renewed';
  
  return this.save();
};

export default mongoose.model('BorrowRecord', borrowRecordSchema);

File: backend-spring/src/main/java/com/library/entity/BorrowRecord.java
package com.library.entity;

import jakarta.persistence.*;
import jakarta.validation.constraints.Max;
import jakarta.validation.constraints.Min;
import jakarta.validation.constraints.NotNull;
import jakarta.validation.constraints.Size;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.LastModifiedDate;
import org.springframework.data.jpa.domain.support.AuditingEntityListener;

import java.math.BigDecimal;
import java.time.LocalDateTime;

@Entity
@Table(name = "borrow_records")
@EntityListeners(AuditingEntityListener.class)
public class BorrowRecord {
    
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;
    
    @ManyToOne(fetch = FetchType.EAGER)
    @JoinColumn(name = "user_id", nullable = false)
    @NotNull(message = "User is required")
    private User user;
    
    @ManyToOne(fetch = FetchType.EAGER)
    @JoinColumn(name = "book_id", nullable = false)
    @NotNull(message = "Book is required")
    private Book book;
    
    @Column(nullable = false)
    private LocalDateTime borrowedAt = LocalDateTime.now();
    
    @Column(nullable = false)
    private LocalDateTime dueAt;
    
    private LocalDateTime returnedAt;
    
    @Min(value = 0, message = "Renewed count cannot be negative")
    @Max(value = 2, message = "Maximum 2 renewals allowed")
    @Column(nullable = false)
    private Integer renewedCount = 0;
    
    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private Status status = Status.BORROWED;
    
    @Min(value = 0, message = "Fine amount cannot be negative")
    @Column(nullable = false, precision = 10, scale = 2)
    private BigDecimal fineAmount = BigDecimal.ZERO;
    
    @Column(nullable = false)
    private Boolean finePaid = false;
    
    @Size(max = 500, message = "Notes cannot exceed 500 characters")
    @Column(length = 500)
    private String notes;
    
    @CreatedDate
    @Column(nullable = false, updatable = false)
    private LocalDateTime createdAt;
    
    @LastModifiedDate
    private LocalDateTime updatedAt;
    
    // Constructors
    public BorrowRecord() {}
    
    public BorrowRecord(User user, Book book) {
        this.user = user;
        this.book = book;
        this.dueAt = LocalDateTime.now().plusDays(14); // Default 14 days
    }
    
    // Business methods
    public boolean isOverdue() {
        return status == Status.BORROWED && LocalDateTime.now().isAfter(dueAt);
    }
    
    public void renew() {
        if (renewedCount >= 2) {
            throw new IllegalStateException("Maximum renewals exceeded");
        }
        if (status != Status.BORROWED && status != Status.RENEWED) {
            throw new IllegalStateException("Can only renew active borrows");
        }
        
        this.dueAt = this.dueAt.plusDays(14);
        this.renewedCount++;
        this.status = Status.RENEWED;
    }
    
    public void returnBook() {
        this.returnedAt = LocalDateTime.now();
        this.status = Status.RETURNED;
        
        // Calculate fine if overdue
        if (returnedAt.isAfter(dueAt)) {
            long daysOverdue = java.time.Duration.between(dueAt, returnedAt).toDays();
            this.fineAmount = BigDecimal.valueOf(daysOverdue * 1.0); // $1 per day
        }
    }
    
    public void updateStatus() {
        if (returnedAt != null) {
            status = Status.RETURNED;
        } else if (LocalDateTime.now().isAfter(dueAt) && status == Status.BORROWED) {
            status = Status.OVERDUE;
            // Calculate fine
            long daysOverdue = java.time.Duration.between(dueAt, LocalDateTime.now()).toDays();
            this.fineAmount = BigDecimal.valueOf(daysOverdue * 1.0);
        }
    }
    
    // Getters and Setters
    public Long getId() {
        return id;
    }
    
    public void setId(Long id) {
        this.id = id;
    }
    
    public User getUser() {
        return user;
    }
    
    public void setUser(User user) {
        this.user = user;
    }
    
    public Book getBook() {
        return book;
    }
    
    public void setBook(Book book) {
        this.book = book;
    }
    
    public LocalDateTime getBorrowedAt() {
        return borrowedAt;
    }
    
    public void setBorrowedAt(LocalDateTime borrowedAt) {
        this.borrowedAt = borrowedAt;
    }
    
    public LocalDateTime getDueAt() {
        return dueAt;
    }
    
    public void setDueAt(LocalDateTime dueAt) {
        this.dueAt = dueAt;
    }
    
    public LocalDateTime getReturnedAt() {
        return returnedAt;
    }
    
    public void setReturnedAt(LocalDateTime returnedAt) {
        this.returnedAt = returnedAt;
    }
    
    public Integer getRenewedCount() {
        return renewedCount;
    }
    
    public void setRenewedCount(Integer renewedCount) {
        this.renewedCount = renewedCount;
    }
    
    public Status getStatus() {
        return status;
    }
    
    public void setStatus(Status status) {
        this.status = status;
    }
    
    public BigDecimal getFineAmount() {
        return fineAmount;
    }
    
    public void setFineAmount(BigDecimal fineAmount) {
        this.fineAmount = fineAmount;
    }
    
    public Boolean getFinePaid() {
        return finePaid;
    }
    
    public void setFinePaid(Boolean finePaid) {
        this.finePaid = finePaid;
    }
    
    public String getNotes() {
        return notes;
    }
    
    public void setNotes(String notes) {
        this.notes = notes;
    }
    
    public LocalDateTime getCreatedAt() {
        return createdAt;
    }
    
    public void setCreatedAt(LocalDateTime createdAt) {
        this.createdAt = createdAt;
    }
    
    public LocalDateTime getUpdatedAt() {
        return updatedAt;
    }
    
    public void setUpdatedAt(LocalDateTime updatedAt) {
        this.updatedAt = updatedAt;
    }
    
    public enum Status {
        BORROWED, RETURNED, OVERDUE, RENEWED
    }
}
